### PR TITLE
NIFI-11532 Remove JUnit 4 and Groovy Test from default dependencies

### PR DESCRIPTION
**Pull request by exceptionfactory**
# Summary

[NIFI-11532](https://issues.apache.org/jira/browse/NIFI-11532) Removes JUnit 4 and Groovy Test modules from the list of default dependencies in the project root configuration, and also removes the Hamcrest dependency from default dependencies.

These changes highlight the extensive work completed to migrate the vast majority of tests to JUnit 5 and Java, leaving a small number of modules and classes on JUnit 4 and Groovy.

Removing JUnit 4 and Groovy Test dependencies from the list of defaults helps avoid introducing JUnit 4 and Groovy code in new modules. A small number of modules with remaining dependencies on JUnit 4 and Groovy have declared dependencies on the JUnit Vintage Engine and the Groovy Test library.

Additional changes include removing several unnecessary test classes from `nifi-socket-utils`, removing duplicative Registry toolkit tests, and removing the integration test classes from the NiFI Kudu module.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-commons/nifi-security-utils-api/src/test/groovy/org/apache/nifi/security/util/TlsConfigurationTest.groovy [hunk 1]
```diff
@@ -1,90 +0,0 @@
-/*
```
Comment (MikeThomsen) [inline nifi-commons/nifi-security-utils-api/src/test/groovy/org/apache/nifi/security/util/TlsConfigurationTest.groovy:1]: Why are you removing a bunch of unit tests without Java replacements in this PR? Did I miss that scoped into the ticket?

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml [hunk 2]
```diff
@@ -65,9 +65,9 @@ language governing permissions and limitations under the License. -->
             <!-- testcontainers requires elements JUnit 4 (even when running in JUnit 5 mode), expected to be removed in 2.x
                 see https://github.com/testcontainers/testcontainers-java/issues/970
              -->
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <scope>compile</scope>
+            <groupId>org.junit.vintage</groupId>
+            <artifactId>junit-vintage-engine</artifactId>
+            <scope>test</scope>
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml:70]: IIRC, that was set as compile because it doesn't come in as a transitive dependency when `nifi-elasticsearch-test-utils` is imported by another project.

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml:70]: If that happens, the tests may silently not run because there's no engine.

File: nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-controller-service/pom.xml [hunk 3]
```diff
@@ -88,12 +88,6 @@
             <artifactId>kudu-client</artifactId>
             <version>${kudu.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.apache.kudu</groupId>
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-controller-service/pom.xml:92]: Why was this removed?

**Review by MikeThomsen — COMMENTED** [review]
Got a few questions.

File: nifi-commons/nifi-security-utils-api/src/test/groovy/org/apache/nifi/security/util/TlsConfigurationTest.groovy (hunk 1 above)
Comment (exceptionfactory) [reply nifi-commons/nifi-security-utils-api/src/test/groovy/org/apache/nifi/security/util/TlsConfigurationTest.groovy:1]: Thanks for the feedback, I put a short explanation for some items in the pull request description, but it is admittedly not described in the Jira issue, so I can certainly adjust the scope one way or the other.

Most of the removals are related directly to JUnit 4 or Groovy Test dependencies, where the tests themselves did not seem worth migrating. If there are some that you think are worth keeping, I can make the adjustments accordingly. To categorize the removals:

1. TLS Configuration and Socket Tests: These tests were limited to checking available TLS versions, primarily for earlier versions of Java 8. With Java 11 as the baseline for the main branch, these tests no longer seem to provide useful checks.

2. Record Utility Tests: These tests were very narrowly scoped with limited methods, and some of the functionality is already covered in other tests

3. Kudu Integration Tests: These tests cover basic uses, but as integration tests, they are not run regularly. [NIFI-11138](https://issues.apache.org/jira/browse/NIFI-11138) outlines potential migration to Testcontainers. These tests require JUnit 4 and since they are narrowly focused, it seems better to rewrite them from scratch using Testcontainers versus building off the existing approach.

4. Registry Properties Tests: With refactoring several versions back, the runtime components are already tested in shared properties classes, so these Registry-specific tests do not add much coverage and even if they were migrated, they should be scaled down significantly.

5. Registry-specific encrypt-config Toolkit Tests: These are Groovy-based tests that also have a lot of overlap with shared encrypt-config code, and shared flow encryption and property encryption modules in nifi-commons.

With that background, I'm open to making adjustments to keep certain tests if you think there is particular value in some cases. I will also update the Jira description to note that some tests should be removed as part of this effort.

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml (hunk 2 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml:70]: Good point, although if that is case, then we should add the dependency to referencing modules, since this module does not appear to have any direction dependencies on JUnit 4. I will look at making some adjustments.

File: nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-controller-service/pom.xml (hunk 3 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-controller-service/pom.xml:92]: See the comments summarizing the removals.

**Comment by MikeThomsen** [comment]
Your reasoning sounds good to me on those points. I'll keep going tomorrow with the review.

**Comment by exceptionfactory** [comment]
> Your reasoning sounds good to me on those points. I'll keep going tomorrow with the review.

Thanks for the reply @MikeThomsen, feel free to follow up with any questions as you proceed. I pushed an update to correct the scope of `junit-vintage-engine` for Elasticsearch integration tests and confirmed that works as expected.

**Comment by exceptionfactory** [comment]
Rebased to address conflicts with recent changes in main branch.

**Comment by exceptionfactory** [comment]
@MikeThomsen I rebased this PR following several recent test updates. Do you have any additional feedback?

**Comment by mattyb149** [comment]
Thanks for this! Merging to main
